feat: Add HyperCache-based authentication

haacked opened the pull request:
Add a new authentication handler `CachedTokenAuthentication` that uses cached access tokens to authenticate requests.

This handler is applied to `local_evaluation` requests, but could be used for any endpoint that requires authentication with a personal access token or a secret api token.

## Problem

When the database is under load or having outages, the `local_evaluation` endpoint starts to throw errors. And this is an important endpoint for customers.

The `local_evaluation` endpoint [already serves up cached flag definitions](https://github.com/PostHog/posthog/pull/36616). But it still requires database calls for authentication. We'd like a solution that's resilient to database outages.

## Changes

- Added signal handlers to handle changes to personal access tokens, users (`is_active`), and teams. The signal handlers build a cache of hashed tokens: personal api keys that have access to the team, `team.secret_api_token` and `team.secret_api_token_backup`
- Added `CachedTokenAuthentication` to hash incoming tokens and look them up in the cache. If they exist, then the user is authenticated.
- If the token is not found in the cache, then authentication falls back to the next `Authentication` class in the list per the standard Django model.
- Schedule warming the cache to run every 10 minutes.

## How did you test this code?

- Bunch of unit tests
- Manual testing.

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Notes

- @benjackwhite and I discussed a Rust service to serve up the cached `local_evaluation`. That's next on my list. But first, we want to make sure that we've made the authentication cache bulletproof. In this implementation, if we missed a cache scenario, authentication falls back to the normal system and everything continues to work.
- This PR is conservative in when to build the cache. For example, we listen to all team changes. We could filter team changes to only rebuild the cache when `api_token`, `secret_api_token`, or `secret_api_token_backup` change. I welcome feedback.
- For users, I worried listening to all changes would cause too much contention so we only listen for changes to `is_active`.

### Cache Format

Cache key format: `cache/team_tokens/{project_api_token}/array/access_tokens.json`
Example: `cache/team_tokens/phc_some-api-token/array/access_tokens.json`

Cache data has three fields:

* `hashed_tokens`: an array of sha256 hashed tokens.
* `team_id`: The team id.
* `last_updated`: when the cache was last updated (we'll use this for etags support later).

Example: 
```json
{
  "hashed_tokens": [
    "sha256$abc123",
    "sha256$def456",
    …
  ],
  "last_updated": "2025-01-15T10:30:00Z",
  "team_id": 42
}
```

## Test Plan

- [ ] Test access with personal access token that has `flags:read` scope for the org.
- [ ] Test access with personal access token that has `flags:write` scope for the org.
- [ ] Test access with personal access token that has `flags:read` scope for all orgs.
- [ ] Test access with personal access token that has `*` scope for the org.
- [ ] Test access with `secure_api_key`
- [ ] Test access with `secure_api_key_backup`
- [ ] Test no access with wrong `secure_api_key`
- [ ] Test no access with wrong `secure_api_key_backup`
- [ ] Test no access with personal access token for the org without `flag:read`
- [ ] Test no access with personal access token with `flags:read` but different org
- [ ] Clear cache and test access
- [ ] Clear cache and test no access
- [ ] Test Organization / project access for a key revoked
- [ ] Test api key rotated / revoked
- [ ] Remove user from organization membership and make sure cache rebuilt
- [ ] Ensure that personal access token last used is updated

haacked commented:
Moving to draft. Feel free to chime in with thoughts. I need to make sure that personal api keys without read access to flags don't authenticate successfully.

haacked commented on posthog/authentication/cached_authentication.py:
```diff
@@ -0,0 +1,329 @@
+"""
+Cached authentication backend for high-traffic API endpoints.
+
+This module provides a Django REST Framework authentication class that uses
+the team access token cache to authenticate requests without database calls.
+"""
```
Not a dumb question. I need to rewrite the problem statement. It isn't so much that auth DB calls are slow, it's that when the database is under load or having other issues, the `local_evaluation` endpoint starts to throw errors. And it's a critical endpoint for our customers.

With this change (and the existing change to cache local evaluation responses), it makes this endpoint resilient to database outages.

Piccirello replied:
Having the DB be online and healthy for authentication, and for `local_evaluation`, seems like a reasonable requirement. I wonder if switching to a DB read replica would help? Or if there's a more direct way we can solve this problem (i.e. make the database more reliable). I'll defer to you on this but caching auth state can get really hairy. State tracking/syncing bugs can lead to incorrect permissions, a session not properly being deleted after a logout, different levels of security management/hardening/access in redis vs our database (do we store sensitive values in Redis today?). This is definitely a problem that falls under the "caching is hard" umbrella.

haacked replied:
@benjackwhite any thoughts on this?

> do we store sensitive values in Redis today?

Unclear. In this implementation, we'll only store the hash of the sensitive values, not the sensitive values themselves.

> a session not properly being deleted after a logout

Not really relevant here. This doesn't do any sort of session tracking.

I see your points though. I think making the database more reliable is a cross-cutting effort. Reducing load on it certainly helps.

I like the idea of a DB read replica. I don't know our current setup for that. I assume we already have a read replica?

benjackwhite replied:
I think @Piccirello , you are over-thinking the range of things we are storing.

We are specifically storing the hash of tokens and api keys so that we can do hyper fast lookups. Loads of systems do this, especially for this kind of high volume use case. The plus sides of being able to offer many 9s for things our customer's SDKs are hitting outweighs the downsides of having to deal with data duplication, cache invalidation etc (which is all managed thanks to the hypercache helpers).

We already have a read replica, we've already thrown money at the problem. This is out well thought through option to finally give the kind of guarantees we want to offer customers. If this was easy, we wouldn't have come up with this solution ;)

benjackwhite replied:
(to be clear i havent fully reviewed yet, only the concept which i believe is sound - we need tests to cover the various cases of scopes changing, keys being deleted etc but otherwise i feel its safe)

haacked replied:
I'll continue with the full implementation and then we can discuss.

Piccirello replied:
Forgive the obvious point, but how long are you intending to cache for? If it's<= 1 hour then it shouldn't interfere with the `last_used_at` tracking logic. Or alternatively could write to a queue to ensure the tracking gets updated later.

haacked replied:
Yeah, I’ll use a background task to do it. We also need to upgrade older tokens. This PR got so big I plan to split it up into the cache warming part and then the actual auth part. But I’ll do the auth part in Rust now that I have a clear picture of everything that needs to happen.

haacked committed:
feat: Add HyperCache-based authentication

Add a new authentication handler that looks for cached access tokens to authenticate requests. Apply this handler to `local_evaluation` requests.

haacked committed:
feat: Add resilient personal API key usage tracking to LocalEvaluationAuthentication

Implement fail-fast Celery task system to track personal API key last_used_at timestamps in LocalEvaluationAuthentication without blocking authentication flows
during database outages.

Key features:
- Fail-fast Celery task with max_retries=0 to prevent out-of-order updates
- Hourly granularity matching existing PersonalAPIKeyAuthentication behavior
- Robust error handling with Prometheus metrics and detailed logging
- Never blocks authentication - all failures logged and ignored
- Always enabled with no configuration required
- Full type safety and comprehensive test coverage

The background task gracefully handles database connectivity issues, missing keys,
and invalid timestamps while maintaining the same update logic as the original
synchronous implementation.

github-actions[bot] committed:
Update query snapshots

haacked commented:
Closing this in favor of https://github.com/PostHog/posthog/pull/37596
